pkg/v1/sdk/capabilities/discovery/cluster_gvr.go

```diff
@@ -35,94 +58,239 @@ func (q *QueryGVR) Name() string {
-// WithVersions needs to be used before calling this function.
+// This method can be omitted to query any resource.
 func (q *QueryGVR) WithResource(resource string) *QueryGVR {
-	q.resource = resource
+	q.resource.set(resource)
 	return q
 }
 
-// Run discovery
+// Run discovery.
 func (q *QueryGVR) Run(config *clusterQueryClientConfig) (bool, error) {
-	if q.group != "" && q.resource != "" && len(q.versions) == 0 {
-		return false, fmt.Errorf("cannot check for group and resource existence without version info; use WithVersion method")
+	if err := q.validate(config); err != nil {
+		return false, fmt.Errorf("validating query: %w", err)
 	}
 
-	var gvrs []schema.GroupVersionResource
-	if len(q.versions) > 0 {
-		for _, v := range q.versions {
-			gvrs = append(gvrs, schema.GroupVersionResource{Group: q.group, Version: v, Resource: q.resource})
+	var unmatched []string
+
+	switch {
+	case q.versions == nil:
+		// q.WithVersions method was omitted and q.versions has not been set.
+		// Any version that matches group and/or resource is considered a match.
+		unmatchedGroupResource, err := q.unmatchedGroupResource(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupResource: %w", err)
+		}
+		if unmatchedGroupResource != "" {
+			unmatched = append(unmatched, unmatchedGroupResource)
+		}
+	case q.resource.String == "":
+		unmatchedGroupVersions, err := q.unmatchedGroupVersions(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupVersions: %w", err)
+		}
+		unmatched = append(unmatched, unmatchedGroupVersions...)
+	case q.resource.String != "":
+		unmatchedGVRs, err := q.unmatchedGroupVersionResources(config)
+		if err != nil {
+			return false, fmt.Errorf("discovering unmatched GroupVersionResources: %w", err)
 		}
-	} else {
-		gvrs = append(gvrs, schema.GroupVersionResource{Group: q.group})
+		unmatched = append(unmatched, unmatchedGVRs...)
+	}
+
+	q.unmatchedGVRs = unmatched
+	return len(unmatched) == 0, nil
+}
+
+func (q *QueryGVR) validate(cfg *clusterQueryClientConfig) error {
+	if cfg == nil {
+		return fmt.Errorf("clusterQueryClientConfig must not be nil")
+	}
+
+	if q.resource.IsSet && q.resource.String == "" {
+		return ErrInvalidWithResourceMethodArgument
+	}
+
+	if q.versions != nil && containsEmpty(q.versions) {
+		return ErrInvalidWithVersionsMethodArgument
+	}
+
+	return nil
+}
+
+// containsEmpty looks for an empty string or items containing only white space(s)
+// in the given slice of strings.
+func containsEmpty(strs []string) bool {
+	for _, s := range strs {
+		s = strings.TrimSpace(s)
+		if s == "" {
+			return true
+		}
+	}
+	return false
+}
+
+func (q *QueryGVR) unmatchedGroupResource(cfg *clusterQueryClientConfig) (string, error) {
+	groups, resources, err := cfg.discoveryClientset.ServerGroupsAndResources()
+	if err != nil {
+		return "", fmt.Errorf("discovering server group and resource: %w", err)
```

Comment:
Same here. This needs to say `failed to discover server group and resource: %w` or remove the wrapping.

Reply:
I would argue that having the word "failed" is redundant. My reasoning is stated in [a previous comment.](https://github.com/vmware-tanzu/tanzu-framework/pull/1724#discussion_r815194795)

Reply:
Per team guidance, I updated the error message to be prefaced with "failed...".